From 86c6c71390e0f66b6ecad1f6999bbbd14865173d Mon Sep 17 00:00:00 2001
From: Avula Sri Charan <quic_asrichar@quicinc.com>
Date: Tue, 21 Nov 2023 18:42:59 +0530
Subject: [PATCH 1/1] wifi: ath12k: ignore ce interrupt until ce pipe init done

CE interrupts are getting triggered immediately after request irq
and before disabling the interrupt, as the srng won't be initialized
by then we are seeing a crash in ISR.

Addressed by adding ce_pipe_init_done flag check and ignoring for
uninitialized srng cases.

Signed-off-by: Avula Sri Charan <quic_asrichar@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/ce.c   | 3 ++-
 drivers/net/wireless/ath/ath12k/core.h | 2 ++
 drivers/net/wireless/ath/ath12k/pci.c  | 4 ++++
 3 files changed, 8 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/ce.c
+++ b/drivers/net/wireless/ath/ath12k/ce.c
@@ -1035,7 +1035,7 @@ int ath12k_ce_init_pipes(struct ath12k_b
 			pipe->status_ring->sw_index = 0;
 		}
 	}
-
+	ab->ce_pipe_init_done = true;
 	return 0;
 }
 
@@ -1082,6 +1082,7 @@ void ath12k_ce_free_pipes(struct ath12k_
 			pipe->status_ring = NULL;
 		}
 	}
+	ab->ce_pipe_init_done = false;
 }
 
 int ath12k_ce_alloc_pipes(struct ath12k_base *ab)
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -1543,6 +1543,8 @@ struct ath12k_base {
 	/* Mode1 specific flag to identify recovering pdev */
 	bool recovery_start;
 	bool rxgainlut_support;
+
+	bool ce_pipe_init_done;
 	/* must be last */
 	u8 drv_priv[0] __aligned(sizeof(void *));
 };
--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -421,6 +421,10 @@ static void ath12k_pci_ce_tasklet(struct
 static irqreturn_t ath12k_pci_ce_interrupt_handler(int irq, void *arg)
 {
 	struct ath12k_ce_pipe *ce_pipe = arg;
+	struct ath12k_base *ab = ce_pipe->ab;
+
+	if (!ab->ce_pipe_init_done)
+		return IRQ_HANDLED;
 
 	/* last interrupt received for this CE */
 	ce_pipe->timestamp = jiffies;
